Add new qnnpack_add and qnnpack_maxpool op to C10 registry

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24103 Add new qnnpack_add and qnnpack_maxpool op to C10 registry**

This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Add new qnnpack_add and qnnpack_maxpool op to C10 registry

This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,145 @@
+    const auto zero_point = input_contig.q_zero_point();
+    qnnp_operator_t qnnpackOperator_{nullptr};
+
+    int64_t padL = padding[0];
```
Maybe rename to `padW` and `padH` to reflect the fact that left/right and top/bottom are symmetrical?

Also, W first when all other args are H first?

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,145 @@
+        "qnnpack_maxpool(): the resulting output Tensor size should be >= 0");
+
+    // NHWC output
+    std::vector<int64_t> outSizes{batch_size, outH, outW, outC};
```
This is a memory allocation, that might be skippable if you pass `{batch_size, outH, outW, outC}` directly as an argument.  Want to try it?

dreiss commented on test/test_quantized.py:
```diff
@@ -737,5 +738,88 @@ def test_qnnpack_linear(self, output_channels, X):
+        scale_B = 5.0
+        zero_point_B = 127
+
+        scale_C = 0.5
```
Maybe vary these?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_add.cpp:
```diff
@@ -0,0 +1,97 @@
+
+    initQNNPACK();
+
+    qnnp_operator_t qnnpackOperator_{nullptr};
```
Yeah, I missed this. Good catch :)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_add.cpp:
```diff
@@ -0,0 +1,97 @@
+
+    size_t volume = qa_contig.numel();
+
+    size_t num_elems = volume / qa_contig.size(0);
```
I think it uses the batch size information to run the op in parallel using the pthreadpool. We can revisit this when we change how QNNPACK uses pthreadpool as well.

dreiss replied:
Okay, but does a 1x10,000 need any different treatment than a 10,000x1 or 100x100 input for pointwise ops?  It seems to me that the total number of elements should be all that matters.

dreiss replied:
Oh, I forgot to notice: this fails with batch size 0.  Can you add a test for that and update the code?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,145 @@
+#endif
+};
+
+static auto registry = torch::RegisterOperators().op(
```
I'm not sure. I don't see anything for the corresponding FBGEMM ops. I think when we add python modules for these ops we can add the relevant docs.

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar commented:
@dreiss addressed all comments. Can you take a quick look?

dreiss reviewed (changes requested):
I'm pushing the QNNPACK changes to support batch size 0.  We should allow that for all of our ops.

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_add.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    const qnnp_status setupStatus = qnnp_setup_add_nc_q8(
+        qnnpack_operator /* add op */,
+        qa_contig.size(0) /* batch size */,
```
this is probably discussed before - do we need care the case that tensor.dim() == 0, i.e. tensor.size() doesn't have element at dim 0?

supriyar replied:
Yes, that is the case of batch_size being 0. I am waiting for @dreiss PR https://github.com/pytorch/pytorch/pull/25044 to land and update this diff to add tests for 0 batch.

ljk53 replied:
batch_size being 0 is qa_contig.size(0) == 0?
I'm concerned about qa_config.size(0) itself is undefined when dim() == 0 (empty tensor without empty sizes tuple)

supriyar replied:
The TORCH_CHECK on line 16 should take care of the case where number of dimensions is 0.

dreiss replied:
I think it's fine for most operators to assert that `dim() > 0`

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_add.cpp:
```diff
@@ -0,0 +1,103 @@
+};
+
+static auto registry = torch::RegisterOperators().op(
+    "quantized::qnnpack_add",
```
do we need full signature? i.e. quantized::qnnpack_add(Tensor qa, Tensor qb, double scale, int zero_point) -> Tensor"? do we need name return tensor? I saw different conventions in the codebase...

supriyar replied:
I'm not sure if there is a convention as I've seen both being used. I felt not having the full signature was better when used for op level logging etc, where it was better to see a concise representation of the function. At the moment we don't have any overloaded functions for qnnpack ops. Do you think I should update this to full signature? @ljk53

ljk53 replied:
Can we improve logging to only print the function name if it's the main concern?

I think @smessmer can comment whether it's some convention we need stick to or not.

smessmer replied:
Logging isn't a concern, you will always have a schema and it will always do the same for logging. If you don't explicitly specify the schema, it'll be inferred from the C++ kernel function, but you'll still have it. We don't have a convention, but as long as you don't use schema features like alias analysis annotations, it's fine to use the shorthand and only specify the name.

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,157 @@
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <aten/src/ATen/native/Pool.h>
```
do we need the "aten/src" part? seems a bit weird to not have it for the other two ATen headers but have it for this one...

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,157 @@
+#ifdef USE_QNNPACK
+  Tensor operator()(
+      Tensor input,
+      std::vector<int64_t> kernel_size,
```
shall we use "torch::List<int64_t>" instead? seems there was an effort to unify it to torch::List but @smessmer might be expert to comment.

smessmer replied:
Yes, please use torch::List.

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,157 @@
+        inC /* input_pixel_stride */,
+        (uint8_t*)qy.data<c10::quint8>() /* output data */,
+        outC /* output_pixel_stride */,
+        nullptr /* thread pool */);
```
what's the purpose of this thread pool param?

supriyar replied:
None that I'm aware of. It is not used in the function at all. caffe2 ops passed in nullptr for this as well.

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_maxpool.cpp:
```diff
@@ -0,0 +1,157 @@
+};
+
+static auto registry = torch::RegisterOperators().op(
+    "quantized::qnnpack_maxpool2d",
```
same here, full function schema signature?

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

supriyar committed:
Update on "Add new qnnpack_add and qnnpack_maxpool op to C10 registry"


This change adds a quantized add and maxpool2d operation for pytorch mobile.

These operators follow the structure of qnnpack in terms of create/setup and run calls. The plan to refactor QNNPACK to make it more functional is currently for FC and Conv ops where the cost of create/setup is high.
For ops like add and maxpool the cost of calling create and setup in each operator invocation is negligible.

Once we migrate FC and Conv QNNPACK ops to be functional in nature, we will consider changing these ops as well to make it consistent.

Differential Revision: [D16734190](https://our.internmc.facebook.com/intern/diff/D16734190/)

ljk53 reviewed (approved):
LGTM - @dreiss wanna take another look?
